tokio/src/util/linked_list/atomic.rs

```diff
@@ -0,0 +1,189 @@
+use super::{Link, LinkedList, LinkedListBase};
+
+use core::cell::UnsafeCell;
+use core::marker::PhantomData;
+use core::mem::ManuallyDrop;
+use core::ptr::NonNull;
+use core::sync::atomic::{
+    AtomicPtr,
+    Ordering::{AcqRel, Relaxed},
+};
+
+/// An atomic intrusive linked list. It allows pushing new nodes concurrently.
+/// Removing nodes still requires an exclusive reference.
+pub(crate) struct AtomicLinkedList<L, T> {
```

Comment:
If you stick with this strategy, I would rename this `ConcurrentPushLinkedList` or something explicit. Also, update the docs to be very clear that synchronization is required for all operations **except** concurrent push.

Reply:
Yeah, I'm inclined to agree with Carl --- the name "AtomicLinkedList" suggests that _all_ links are atomic...

Reply:
also, it might be worth investigating whether there are performance benefits to using this list type in other places in `tokio::sync`. i  think there are other synchronization primitives that currently force all tasks pushing to their wait lists pushes to serialize themselves with a `Mutex`, which could potentially benefit from this type. but, i would save that for a separate branch.

Reply:
> also, it might be worth investigating whether there are performance benefits to using this list type in other places in `tokio::sync`.

I was eyeing `sync::Notify`: it has a mutex that guards a list of waiters. #5464 reduced contention on the watch channel by utilizing 8 instances of `Notify` instead of one. Looks like the perfect candidate.

Reply:
Yes, there are definitely other places that could benefit from something similar. I'm also wondering if the timer could have a similar optimization. Some people experience significant contention in the timer.

However, timers have more contention on cancellation due to timeouts usually being cancelled...